x-pack/plugins/ml/public/application/services/dashboard_service.ts

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObjectsClientContract } from 'kibana/public';
+import { htmlIdGenerator } from '@elastic/eui';
+import { useMemo } from 'react';
+import {
+  DASHBOARD_APP_URL_GENERATOR,
+  DashboardUrlGenerator,
+  SavedDashboardPanel,
+  SavedObjectDashboard,
+} from '../../../../../../src/plugins/dashboard/public';
+import { useMlKibana } from '../contexts/kibana';
+
+export type DashboardService = ReturnType<typeof dashboardServiceProvider>;
+
+export function dashboardServiceProvider(
```

Comment:
@majagrubic, @ThomThomson, I think you will be interested in reviewing this file :) 

Also please see my high-level understanding of the flow, maybe you will have other ideas: 

### As I understand the flow now:

1. User in ML does the action
2. ML code fetches dashboard saved objects, adds a panel, saves it back 
3. ML navigates to just updated dashboard

### How it should be improved, so ml doesn't manipulate dashboard SO directly

dashboard plugin should expose this dashboard service and ml will use it 

1. User in ML does the action
2. ML code uses dashboard service from dashboard plugin, which fetches the SO, adds a panel, saves it back 
3. ML navigates to just updated dashboard

### Making UX better and consistent with visualize & lens

Currently ML changes dashboard without user's explicit consent to save it. 
But when saving visualisation in visualise, dashboard isn't actually changed until user presses "Save" in a dashboard. 

Probably we could follow the same approach as in visualize and forward new panel in location.state after this is merged: https://github.com/elastic/kibana/pull/67064 @ThomThomson . The difference is that now it supports forwarding SO id, but here we need to forward panel "byValue" 
So we need to extend possible incoming state here: https://github.com/elastic/kibana/pull/67064/files#diff-1cfae33faae54944efc865822545348fR433

---

We discussed this with @elastic/kibana-app-arch, we think current approach could be merged as is because we don't have existing alternative, but would be nice if we create a ticket for follow up from dashboard perspective and referenced it here from code as TODO

Reply:
Created an issue https://github.com/elastic/kibana/issues/69152 and added TODO comment in https://github.com/elastic/kibana/pull/68784/commits/1c3e127d7b1024f01f944f47216bb629930bc417 to update after kibana issue is resolved.

Reply:
Good call! @majagrubic is making great progress on adding visualization panels by value, and will add a new `embeddableByValue` package to the state transfer system after #67064 is merged. Just started looking at this PR, but it definitely seems like we could change this to use that method later.

Reply:
@ThomThomson, sounds great! then bookmark https://github.com/elastic/kibana/issues/69152 ;)